### PR TITLE
Rework Schema to have separate lists for account and email for whitel…

### PR DESCRIPTION
**Pull request by CodeWolf86**
…isting purposes.

### CODE REVIEW

File: src/SFA.DAS.Authorization.EmployerFeatures/Context/AuthorizationContextExtensions.cs [hunk 1]
```diff
@@ -4,10 +4,17 @@ namespace SFA.DAS.Authorization.EmployerFeatures.Context
...
     {
-        public static void AddEmployerFeatureValues(this IAuthorizationContext authorizationContext, long accountId, string userEmail)
+        public static void AddEmployerFeatureValues(this IAuthorizationContext authorizationContext, long? accountId, string userEmail)
         {
-            authorizationContext.Set(AuthorizationContextKey.AccountId, accountId);
-            authorizationContext.Set(AuthorizationContextKey.UserEmail, userEmail);
+            if (accountId.HasValue)
+            {
+                authorizationContext.Set(AuthorizationContextKey.AccountId, accountId);
+            }
```
Comment (Pauljgraham) [inline src/SFA.DAS.Authorization.EmployerFeatures/Context/AuthorizationContextExtensions.cs]: From memory, there is a slight difference between the AuthorizationContext returning a null value and it not finding a value. I'm not sure that will make a difference to the consuming code.

File: src/SFA.DAS.Authorization.EmployerFeatures/Context/AuthorizationContextExtensions.cs [hunk 2]
```diff
@@ -4,10 +4,17 @@ namespace SFA.DAS.Authorization.EmployerFeatures.Context
...
+
+            if (!string.IsNullOrWhiteSpace(userEmail))
+            {
+                authorizationContext.Set(AuthorizationContextKey.UserEmail, userEmail);
+            }
```
Comment (Pauljgraham) [inline src/SFA.DAS.Authorization.EmployerFeatures/Context/AuthorizationContextExtensions.cs]: Same concern as above

File: src/SFA.DAS.Authorization.EmployerFeatures/Models/EmployerFeatureToggle.cs [hunk 3]
```diff
@@ -1,26 +1,26 @@
...
 
         public bool IsUserWhitelisted(long? accountId, string userEmail)
         {
-            return accountId.HasValue
-                ? Whitelist.Any(w => (w.AccountId == accountId || w.AccountId == null) && IsUserWhitelisted(userEmail))
-                : IsUserWhitelisted(userEmail);
-        }
+            return (accountId.HasValue && AccountWhitelist.Contains(accountId.Value))
+                ||
+                (!string.IsNullOrWhiteSpace(userEmail) && EmailWhitelist.Any(email => Regex.IsMatch(userEmail, email, RegexOptions.IgnoreCase)));
```
Comment (Pauljgraham) [inline src/SFA.DAS.Authorization.EmployerFeatures/Models/EmployerFeatureToggle.cs]: As a matter of interest is Regex.IsMatch() quicker that than the c# Equals() & OrdinalIgnoreCase. Also possible to get a partial match (ie will A@AAA.COM match AAA@AAA.COM)

or could you use EmailWhiteList.Contains(userEmail, StringComparer.InvariantCultureIgnoreCase)

I'm not sure which option is optimal

Comment (CodeWolf86) [reply src/SFA.DAS.Authorization.EmployerFeatures/Models/EmployerFeatureToggle.cs]: So some improvements could be made here. The switch to Regex wasn't performance based. The Das Auth implementation in MA allowed for wildcarding, so some of the config is set as `"EmailWhitelist":[".*@eoi.com","nlapptest2@mailinator.com"]}` 

Where a .Contains doesn't satisfy that criteria of matching any eoi.com domain email address. If there are concerns over speed, we could check a .Contains first in an attempt to short circuit and fall back on the Regex. Currently in live there are only about 26 email addresses in the config, and no wildcarding. So happy to amend this if we think there is an issue.

A partial match I hadn't considered, I'll write a unit test for that and maybe see if I can change it to be stricter where an email address is concerned.

Comment (CodeWolf86) [reply src/SFA.DAS.Authorization.EmployerFeatures/Models/EmployerFeatureToggle.cs]: I've updated with a second commit to hopefully address the partial match and to also leverage the string comparer and equals for exact matches.

**Pull request merged**